lms/djangoapps/instructor_task/tests/test_tasks_helper.py

```diff
@@ -462,25 +462,101 @@ def test_team_deleted(self):
         self._verify_cell_data_for_user(self.student2.username, self.course.id, 'Team Name', team2.name)
 
 
-class TestProblemResponsesReport(TestReportMixin, InstructorTaskCourseTestCase):
+# pylint: disable=protected-access
+class TestProblemResponsesReport(TestReportMixin, InstructorTaskModuleTestCase):
     """
     Tests that generation of CSV files listing student answers to a
     given problem works.
     """
     def setUp(self):
         super(TestProblemResponsesReport, self).setUp()
-        self.course = CourseFactory.create()
+        self.initialize_course()
+        self.instructor = self.create_instructor('instructor')
+        self.student = self.create_student('student')
+
+    # Can be used once CapaDescriptor gets ``generate_report_data`` method.
+    # @contextmanager
+    # def _remove_report_generator(self):
+    #     """
+    #     Temporarily removes the generate_report_data method so we can test
+    #     report generation when it's absent.
+    #     """
+    #     from xmodule.capa_module import CapaDescriptor
+    #     generate_report_data = CapaDescriptor.generate_report_data
+    #     del CapaDescriptor.generate_report_data
+    #     yield
+    #     CapaDescriptor.generate_report_data = generate_report_data
+
+    @patch.dict("django.conf.settings.FEATURES", {"MAX_PROBLEM_RESPONSES_COUNT": 4})
+    def test_build_student_data_limit(self):
+        self.define_option_problem(u'Problem1')
+        for ctr in range(5):
+            student = self.create_student('student{}'.format(ctr))
+            self.submit_student_answer(student.username, u'Problem1', ['Option 1'])
+
+        student_data = ProblemResponses._build_student_data(
+            user_id=self.instructor.id,
+            course_key=self.course.id,
+            usage_key_str=str(self.course.location),
+        )
+
+        self.assertEquals(len(student_data), 4)
+
+    def test_build_student_data_no_generate(self):
+        self.define_option_problem(u'Problem1')
+        self.submit_student_answer(self.student.username, u'Problem1', ['Option 1'])
+
+        student_data = ProblemResponses._build_student_data(
+            user_id=self.instructor.id,
+            course_key=self.course.id,
+            usage_key_str=str(self.course.location),
+        )
+        self.assertEquals(len(student_data), 1)
+        self.assertDictContainsSubset({
+            'username': 'student',
+            'location': 'test_course > Section > Subsection > Problem1',
+            'block_key': 'i4x://edx/1.23x/problem/Problem1',
+            'title': 'Problem1',
+        }, student_data[0])
+        self.assertIn('state', student_data[0])
```

Comment:
Can we not verify the exact value of `state`? If not, another thing that can be checked is if `list_problem_responses()` is called. The point being, if someone were add `generate_report_data` to whatever block type is being used in this test, the test will continue to pass even though it will not be taking the code path it is taking right now.

Reply:
On second thoughts, since the other asserts in these two tests are the same, you should assert that `list_problem_responses()` is called in this test and not in the next test.

Here is an example of how to set it up: https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/tests/test_email.py#L529-L533

Reply:
Sounds good.